lib/Doctrine/DBAL/Driver/PDOPgSql/Driver.php

```diff
@@ -53,6 +53,14 @@ public function connect(array $params, $username = null, $password = null, array
                 $pdo->setAttribute(PDO::PGSQL_ATTR_DISABLE_PREPARES, true);
             }
 
+            /* defining client_encoding via SET NAMES to avoid inconsistent DSN support
+             * - the 'client_encoding' connection param only works with postgres >= 9.1
+             * - passing client_encoding via the 'options' param breaks pgbouncer support
+             */
+            if (isset($params['charset'])) {
+              $pdo->query('SET NAMES \''.$params['charset'].'\'');
```

Comment:
@rocksfrow @deeky666 we could just use the 9.1+ method of connecting inside a try block - if it fails, check the message for 'client_encoding' and connect again using the fallback method with SET NAMES.

We could also check first if the pgsql extension is loaded -- pgsql has a constant PGSQL_LIBPQ_VERSION which tells what client version is in use. If PGSQL_LIBPQ_VERSION >= 9.1, use client_encoding in the dsn. If not, use the fallback method. 

This would be really easy to fix if PDO would allow us to get more info about drivers...

Another alternative would be to run phpinfo and find "PostgreSQL\(libpq\) Version => ([0-9\.]*)"... But that's just dumb.